File: datastage-server/src/utils/cache.ts
/**
 * Simple in-memory cache implementation
 */

import { CacheEntry } from '../config/types.js';
import { logger } from './logger.js';

export class SimpleCache {
  private cache: Map<string, CacheEntry<any>>;

  constructor() {
    this.cache = new Map();
  }

  /**
   * Get a value from cache
   */
  get<T>(key: string): T | null {
    const entry = this.cache.get(key);
    
    if (!entry) {
      logger.debug(`Cache miss for key: ${key}`);
      return null;
    }

    // Check if entry has expired
    const now = Date.now();
    if (now > entry.timestamp + entry.ttl * 1000) {
      logger.debug(`Cache entry expired for key: ${key}`);
      this.cache.delete(key);
      return null;
    }

    logger.debug(`Cache hit for key: ${key}`);
    return entry.data as T;
  }

  /**
   * Set a value in cache
   */
  set<T>(key: string, data: T, ttl: number): void {
    const entry: CacheEntry<T> = {
      data,
      timestamp: Date.now(),
      ttl,
    };

    this.cache.set(key, entry);
    logger.debug(`Cached data for key: ${key} (TTL: ${ttl}s)`);
  }

  /**
   * Delete a value from cache
   */
  delete(key: string): boolean {
    const deleted = this.cache.delete(key);
    if (deleted) {
      logger.debug(`Deleted cache entry for key: ${key}`);
    }
    return deleted;
  }

  /**
   * Clear all cache entries
   */
  clear(): void {
    const size = this.cache.size;
    this.cache.clear();
    logger.debug(`Cleared ${size} cache entries`);
  }

  /**
   * Get cache size
   */
  size(): number {
    return this.cache.size;
  }

  /**
   * Check if key exists in cache
   */
  has(key: string): boolean {
    const entry = this.cache.get(key);
    if (!entry) {
      return false;
    }

    // Check if expired
    const now = Date.now();
    if (now > entry.timestamp + entry.ttl * 1000) {
      this.cache.delete(key);
      return false;
    }

    return true;
  }

  /**
   * Clean up expired entries
   */
  cleanup(): void {
    const now = Date.now();
    let cleaned = 0;

    for (const [key, entry] of this.cache.entries()) {
      if (now > entry.timestamp + entry.ttl * 1000) {
        this.cache.delete(key);
        cleaned++;
      }
    }

    if (cleaned > 0) {
      logger.debug(`Cleaned up ${cleaned} expired cache entries`);
    }
  }
}

// Made with Bob

File: python-mcp/src/utils/cache.py
"""
Simple in-memory cache implementation
"""

import time
from typing import Any, Optional, Dict
from ..utils.logger import logger


class CacheEntry:
    """Cache entry with data, timestamp, and TTL"""
    def __init__(self, data: Any, ttl: float):
        self.data = data
        self.timestamp = time.time()
        self.ttl = ttl


class SimpleCache:
    """Simple in-memory cache"""
    
    def __init__(self):
        self.cache: Dict[str, CacheEntry] = {}
    
    def get(self, key: str) -> Optional[Any]:
        """Get a value from cache"""
        entry = self.cache.get(key)
        
        if not entry:
            logger.debug(f'Cache miss for key: {key}')
            return None
        
        # Check if entry has expired
        now = time.time()
        if now > entry.timestamp + entry.ttl:
            logger.debug(f'Cache entry expired for key: {key}')
            del self.cache[key]
            return None
        
        logger.debug(f'Cache hit for key: {key}')
        return entry.data
    
    def set(self, key: str, data: Any, ttl: float) -> None:
        """Set a value in cache"""
        entry = CacheEntry(data, ttl)
        self.cache[key] = entry
        logger.debug(f'Cached data for key: {key} (TTL: {ttl}s)')
    
    def delete(self, key: str) -> bool:
        """Delete a value from cache"""
        if key in self.cache:
            del self.cache[key]
            logger.debug(f'Deleted cache entry for key: {key}')
            return True
        return False
    
    def clear(self) -> None:
        """Clear all cache entries"""
        size = len(self.cache)
        self.cache.clear()
        logger.debug(f'Cleared {size} cache entries')
    
    def size(self) -> int:
        """Get cache size"""
        return len(self.cache)
    
    def has(self, key: str) -> bool:
        """Check if key exists in cache"""
        entry = self.cache.get(key)
        if not entry:
            return False
        
        # Check if expired
        now = time.time()
        if now > entry.timestamp + entry.ttl:
            del self.cache[key]
            return False
        
        return True
    
    def cleanup(self) -> None:
        """Clean up expired entries"""
        now = time.time()
        cleaned = 0
        
        keys_to_delete = []
        for key, entry in self.cache.items():
            if now > entry.timestamp + entry.ttl:
                keys_to_delete.append(key)
                cleaned += 1
        
        for key in keys_to_delete:
            del self.cache[key]
        
        if cleaned > 0:
            logger.debug(f'Cleaned up {cleaned} expired cache entries')


# Made with Bob